pkg/controllers/machinesync/machine_sync_controller.go

```diff
@@ -99,7 +111,80 @@ func (r *MachineSyncReconciler) SetupWithManager(mgr ctrl.Manager) error {
 }
 
 // Reconcile reconciles CAPI and MAPI machines for their respective namespaces.
+//
+//nolint:funlen
 func (r *MachineSyncReconciler) Reconcile(ctx context.Context, req reconcile.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx, "namespace", req.Namespace, "name", req.Name)
+
+	logger.V(1).Info("Reconciling machine")
+	defer logger.V(1).Info("Finished reconciling machine")
+
+	var (
+		mapiMachineNotFound, capiMachineNotFound bool
+	)
+
+	// Get the MAPI Machine.
+	mapiMachine := &machinev1beta1.Machine{}
+	mapiNamespacedName := client.ObjectKey{
+		Namespace: r.MAPINamespace,
+		Name:      req.Name,
+	}
+
+	if err := r.Get(ctx, mapiNamespacedName, mapiMachine); apierrors.IsNotFound(err) {
+		logger.Info("MAPI Machine not found")
+
+		mapiMachineNotFound = true
+	} else if err != nil {
+		logger.Error(err, "Failed to get MAPI Machine")
+		return ctrl.Result{}, fmt.Errorf("%w: %w", errFailedToGetMAPIMachine, err)
+	}
+
+	// Get the corresponding CAPI Machine.
+	capiMachine := &capiv1beta1.Machine{}
+	capiNamespacedName := client.ObjectKey{
+		Namespace: r.CAPINamespace,
+		Name:      req.Name,
+	}
+
+	if err := r.Get(ctx, capiNamespacedName, capiMachine); apierrors.IsNotFound(err) {
+		logger.Info("CAPI Machine not found")
+
+		capiMachineNotFound = true
+	} else if err != nil {
+		logger.Error(err, "Failed to get CAPI Machine")
+		return ctrl.Result{}, fmt.Errorf("%w: %w", errFailedToGetCAPIMachine, err)
+	}
+
+	if mapiMachineNotFound && capiMachineNotFound {
+		logger.Info("CAPI and MAPI machines not found, nothing to do")
+		return ctrl.Result{}, nil
+	}
+
+	// We mirror if the CAPI machine is owned by a MachineSet which has a MAPI
+	// counterpart. This is because we want to be able to migrate in both directions.
+	if mapiMachineNotFound {
+		if shouldReconcile, err := r.shouldMirrorCAPIMachineToMAPIMachine(ctx, capiMachine, logger); err != nil {
+			return ctrl.Result{}, err
+		} else if shouldReconcile {
+			return r.reconcileCAPIMachinetoMAPIMachine(ctx, capiMachine, mapiMachine)
+		}
+	}
+
+	// If we have not set AuthoritativeAPI, we default to MachineAuthorityMachineAPI
+	if mapiMachine.Spec.AuthoritativeAPI == machinev1beta1.MachineAuthorityClusterAPI {
+		return r.reconcileCAPIMachinetoMAPIMachine(ctx, capiMachine, mapiMachine)
+	} else {
+		return r.reconcileMAPIMachinetoCAPIMachine(ctx, mapiMachine, capiMachine)
+	}
```

Comment:
Would probably prefer a switch here that explicitly checks the empty string, MAPI and CAPI string values, and then errors for an unknown value, else you're defaulting unknown values to a MAPI to CAPI direction here.

Is that intended?

Reply:
>Is that intended?

Going from the comment, I was thinking so... but can't for the life of me remember why now. I think possibly I recall a webhook going in to ensure it's never unset, however that still doesn't mean this would make sense 🤔 

Either way, will update to a switch!